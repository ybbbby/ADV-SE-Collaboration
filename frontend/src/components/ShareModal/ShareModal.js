import React from 'react'
import PropTypes from 'prop-types'
import {
  Button,
  InputBase,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  IconButton,
  Paper,
  Tooltip,
  Link,
  ClickAwayListener,
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import FileCopyIcon from '@material-ui/icons/FileCopy'
import TwitterIcon from '@material-ui/icons/Twitter'
import FacebookIcon from '@material-ui/icons/Facebook'
import EmailIcon from '@material-ui/icons/Email'
import { CopyToClipboard } from 'react-copy-to-clipboard'

const useStyles = makeStyles((theme) => ({
  input: {
    marginLeft: theme.spacing(1),
    flex: 1,
  },
  iconButton: {
    padding: 10,
  },
  root: {
    padding: '2px 4px',
    display: 'flex',
    alignItems: 'center',
    width: '100%',
  },
  socialMedia: {
    minWidth: '500px',
    textAlign: 'center',
  },
  socialIcon: {
    margin: '10px',
    height: '30px',
    width: '30px',
  },
}))

const TWITTER_SHARE_LINK = 'https://twitter.com/intent/tweet?url='
const FACEBOOK_SHARE_LINK = 'https://www.facebook.com/sharer/sharer.php?u='
const EMAIL_SHARE_LINK = 'mailto:info@example.com?&subject=&body='

function ShareModal(props) {
  const classes = useStyles()
  const { url, handleClose, open } = props
  const [copied, setCopied] = React.useState(false)

  return (
    <Dialog
      open={open}
      onClose={() => handleClose()}
      aria-labelledby="form-dialog-title"
    >
      <DialogTitle id="form-dialog-title">Share the event</DialogTitle>
      <DialogContent>
        <div className={classes.socialMedia}>
          <Tooltip title="Share on Twitter">
            <Link
              href={TWITTER_SHARE_LINK + url}
              color="inherit"
              target="_blank"
            >
              <TwitterIcon className={classes.socialIcon} />
            </Link>
          </Tooltip>
          <Tooltip title="Share on Facebook">
            <Link
              href={FACEBOOK_SHARE_LINK + url}
              color="inherit"
              target="_blank"
            >
              <FacebookIcon className={classes.socialIcon} />
            </Link>
          </Tooltip>
          <Tooltip title="Share by Email">
            <Link href={EMAIL_SHARE_LINK + url} color="inherit">
              <EmailIcon className={classes.socialIcon} />
            </Link>
          </Tooltip>
        </div>
        <Paper className={classes.root}>
          <InputBase
            className={classes.input}
            placeholder={url}
            disabled
            inputProps={{ 'aria-label': 'event sharing link' }}
          />
          <ClickAwayListener onClickAway={() => setCopied(false)}>
            <CopyToClipboard text={url} onCopy={() => setCopied(true)}>
              <Tooltip
                PopperProps={{
                  disablePortal: true,
                }}
                open={copied}
                disableFocusListener
                disableHoverListener
                title="Copied"
                arrow
              >
                <IconButton className={classes.iconButton} aria-label="copy">
                  <FileCopyIcon />
                </IconButton>
              </Tooltip>
            </CopyToClipboard>
          </ClickAwayListener>
        </Paper>
      </DialogContent>
      <DialogActions>
        <Button onClick={() => handleClose()} color="primary">
          Cancel
        </Button>
      </DialogActions>
    </Dialog>
  )
}

ShareModal.propTypes = {
  open: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  url: PropTypes.string.isRequired,
}

export default ShareModal
