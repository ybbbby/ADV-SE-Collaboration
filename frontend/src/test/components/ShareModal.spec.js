/* eslint-env jest */
import { shallow } from 'enzyme'
import React from 'react'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import ShareModal from '../../components/ShareModal/ShareModal'

const url = 'www.google.com'

describe('ShareModal', () => {
  test('does not render modal when closed', () => {
    let openShare = false
    const close = jest.fn(() => (openShare = false))
    render(
      <ShareModal handleClose={() => close()} open={openShare} url={url} />
    )
    expect(screen.queryByText('Welcome Back!')).toBeNull()
  })

  test('renders modal when opened', () => {
    let openShare = true
    const close = jest.fn(() => (openShare = false))
    render(
      <ShareModal handleClose={() => close()} open={openShare} url={url} />
    )
    expect(screen.queryByText(url)).toBeDefined()
  })

  test('click copy button', async () => {
    const jsdomPrompt = window.prompt
    window.prompt = () => {}
    let openShare = true
    const close = jest.fn(() => (openShare = false))
    render(
      <ShareModal handleClose={() => close()} open={openShare} url={url} />
    )
    await waitFor(() =>
      userEvent.click(screen.getByRole('button', { name: 'copy' }))
    )
    expect(screen.queryByText('Copied')).toBeDefined()
    window.prompt = jsdomPrompt
  })

  test('close modal', async () => {
    let openShare = true
    const close = jest.fn(() => (openShare = false))
    render(
      <ShareModal handleClose={() => close()} open={openShare} url={url} />
    )
    await waitFor(() =>
      userEvent.click(screen.getByRole('button', { name: 'Cancel' }))
    )
    expect(close).toHaveBeenCalledTimes(1)
  })

  test('close modal by clicking away', () => {
    let openShare = true
    const close = jest.fn(() => (openShare = false))
    render(
      <ShareModal handleClose={() => close()} open={openShare} url={url} />
    )
    fireEvent.keyDown(screen.getByRole('presentation'), {
      key: 'Escape',
      code: 'Escape',
    })
    expect(close).toHaveBeenCalledTimes(1)
  })

  it('snapshot when open', () => {
    let openShare = true
    const close = jest.fn(() => (openShare = false))
    const wrapper = shallow(
      <ShareModal handleClose={() => close()} open={openShare} url={url} />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('snapshot when closed', () => {
    let openShare = false
    const close = jest.fn(() => (openShare = false))
    const wrapper = shallow(
      <ShareModal handleClose={() => close()} open={openShare} url={url} />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
