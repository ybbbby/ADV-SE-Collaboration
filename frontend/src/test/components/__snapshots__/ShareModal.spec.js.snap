// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareModal snapshot when closed 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-labelledby="form-dialog-title"
  onClose={[Function]}
  open={false}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="form-dialog-title"
  >
    Share the event
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))>
    <div
      className="makeStyles-socialMedia-29"
    >
      <WithStyles(ForwardRef(Tooltip))
        title="Share on Twitter"
      >
        <WithStyles(ForwardRef(Link))
          color="inherit"
          href="https://twitter.com/intent/tweet?url=www.google.com"
          target="_blank"
        >
          <Memo([object Object])
            className="makeStyles-socialIcon-30"
          />
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Tooltip))>
      <WithStyles(ForwardRef(Tooltip))
        title="Share on Facebook"
      >
        <WithStyles(ForwardRef(Link))
          color="inherit"
          href="https://www.facebook.com/sharer/sharer.php?u=www.google.com"
          target="_blank"
        >
          <Memo([object Object])
            className="makeStyles-socialIcon-30"
          />
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Tooltip))>
      <WithStyles(ForwardRef(Tooltip))
        title="Share by Email"
      >
        <WithStyles(ForwardRef(Link))
          color="inherit"
          href="mailto:info@example.com?&subject=&body=www.google.com"
        >
          <Memo([object Object])
            className="makeStyles-socialIcon-30"
          />
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Tooltip))>
    </div>
    <WithStyles(ForwardRef(Paper))
      className="makeStyles-root-28"
    >
      <WithStyles(ForwardRef(InputBase))
        className="makeStyles-input-26"
        inputProps={
          Object {
            "aria-label": "event sharing link",
            "readOnly": true,
          }
        }
        placeholder="www.google.com"
      />
      <ClickAwayListener
        onClickAway={[Function]}
      >
        <CopyToClipboard
          onCopy={[Function]}
          text="www.google.com"
        >
          <WithStyles(ForwardRef(Tooltip))
            PopperProps={
              Object {
                "disablePortal": true,
              }
            }
            arrow={true}
            disableFocusListener={true}
            disableHoverListener={true}
            open={false}
            title="Copied"
          >
            <WithStyles(ForwardRef(IconButton))
              aria-label="copy"
              className="makeStyles-iconButton-27"
            >
              <Memo([object Object]) />
            </WithStyles(ForwardRef(IconButton))>
          </WithStyles(ForwardRef(Tooltip))>
        </CopyToClipboard>
      </ClickAwayListener>
    </WithStyles(ForwardRef(Paper))>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`ShareModal snapshot when open 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-labelledby="form-dialog-title"
  onClose={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="form-dialog-title"
  >
    Share the event
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))>
    <div
      className="makeStyles-socialMedia-29"
    >
      <WithStyles(ForwardRef(Tooltip))
        title="Share on Twitter"
      >
        <WithStyles(ForwardRef(Link))
          color="inherit"
          href="https://twitter.com/intent/tweet?url=www.google.com"
          target="_blank"
        >
          <Memo([object Object])
            className="makeStyles-socialIcon-30"
          />
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Tooltip))>
      <WithStyles(ForwardRef(Tooltip))
        title="Share on Facebook"
      >
        <WithStyles(ForwardRef(Link))
          color="inherit"
          href="https://www.facebook.com/sharer/sharer.php?u=www.google.com"
          target="_blank"
        >
          <Memo([object Object])
            className="makeStyles-socialIcon-30"
          />
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Tooltip))>
      <WithStyles(ForwardRef(Tooltip))
        title="Share by Email"
      >
        <WithStyles(ForwardRef(Link))
          color="inherit"
          href="mailto:info@example.com?&subject=&body=www.google.com"
        >
          <Memo([object Object])
            className="makeStyles-socialIcon-30"
          />
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Tooltip))>
    </div>
    <WithStyles(ForwardRef(Paper))
      className="makeStyles-root-28"
    >
      <WithStyles(ForwardRef(InputBase))
        className="makeStyles-input-26"
        inputProps={
          Object {
            "aria-label": "event sharing link",
            "readOnly": true,
          }
        }
        placeholder="www.google.com"
      />
      <ClickAwayListener
        onClickAway={[Function]}
      >
        <CopyToClipboard
          onCopy={[Function]}
          text="www.google.com"
        >
          <WithStyles(ForwardRef(Tooltip))
            PopperProps={
              Object {
                "disablePortal": true,
              }
            }
            arrow={true}
            disableFocusListener={true}
            disableHoverListener={true}
            open={false}
            title="Copied"
          >
            <WithStyles(ForwardRef(IconButton))
              aria-label="copy"
              className="makeStyles-iconButton-27"
            >
              <Memo([object Object]) />
            </WithStyles(ForwardRef(IconButton))>
          </WithStyles(ForwardRef(Tooltip))>
        </CopyToClipboard>
      </ClickAwayListener>
    </WithStyles(ForwardRef(Paper))>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;
